background worker "Influx line protocol listener" was terminated by signal 11: Segmentation fault

Hi, I was trying to install pg_influx but I get a Segmentation fault

When I call worker_launch, the worker is created successfully:
```
postgres=# select pg_sleep(1) from worker_launch('db_worker', 4711::text);
NOTICE:  background worker started
DETAIL:  pid=832
 pg_sleep 
----------
 
(1 row)
```

But after a few seconds, postgres freezes and kills the worker process:
```
2022-09-13 19:24:29.344 UTC [832] LOG:  worker initialized
2022-09-13 19:24:29.344 UTC [832] DETAIL:  service=4711, database_id=13457, namespace_id=34136
---> freezes here!
2022-09-13 19:25:30.516 UTC [1] LOG:  background worker "Influx line protocol listener" (PID 832) was terminated by signal 11: Segmentation fault
2022-09-13 19:25:30.516 UTC [1] DETAIL:  Failed process was running: initializing worker
2022-09-13 19:25:30.516 UTC [1] LOG:  terminating any other active server processes
2022-09-13 19:25:30.517 UTC [822] WARNING:  terminating connection because of crash of another server process
2022-09-13 19:25:30.517 UTC [822] DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2022-09-13 19:25:30.517 UTC [822] HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2022-09-13 19:25:30.517 UTC [613] WARNING:  terminating connection because of crash of another server process
2022-09-13 19:25:30.517 UTC [613] DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2022-09-13 19:25:30.517 UTC [613] HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2022-09-13 19:25:30.532 UTC [1] LOG:  all server processes terminated; reinitializing
2022-09-13 19:25:32.390 UTC [838] LOG:  database system was interrupted; last known up at 2022-09-13 19:02:43 UTC
2022-09-13 19:25:32.431 UTC [838] LOG:  database system was not properly shut down; automatic recovery in progress
2022-09-13 19:25:32.436 UTC [838] LOG:  redo starts at 0/1A4F968
2022-09-13 19:25:32.437 UTC [838] LOG:  invalid record length at 0/1A50628: wanted 24, got 0
2022-09-13 19:25:32.437 UTC [838] LOG:  redo done at 0/1A505F0
2022-09-13 19:25:32.482 UTC [1] LOG:  database system is ready to accept connections
2022-09-13 19:25:32.488 UTC [844] LOG:  TimescaleDB background worker launcher connected to shared catalogs
```

I discovered that the function that is raising the segmentation fault is [PushActiveSnapshot](https://github.com/timescale/pg_influx/blob/main/worker.c#L275) but it is a Postgres function.

After reading the post [The Server Programming Interface](https://dbmsdrops.kindahl.net/2022/04/23/the-server-programming-interface/) I found that the use of the function is necessary for the snapshot set up. However, I don't understand why is this causing the segfault. 

I found this [issue](https://www.mail-archive.com/search?l=pgsql-hackers@lists.postgresql.org&q=subject:%22Re%5C%3A+Segment+fault+when+excuting+SPI+function+On+PG+with+commit+41c6a5be%22&o=newest&f=1) that refers to the Postgres [41c6a5be](https://github.com/postgres/postgres/commit/41c6a5be) commit for the [REL_12_STABLE](https://github.com/postgres/postgres/compare/REL_12_STABLE) and it seems to be a possble cause of the segfault. 
I want to know if you had any issue with newer postgres versions (major than 12) as its happened to me or if this code was only tested on older postgres versions

Btw, I can't find any other reference or documentation about the function (except in the [postgres source code](https://github.com/postgres/postgres/blob/REL_13_STABLE/src/backend/utils/time/snapmgr.c#L737) or [postgres spi documentation](https://www.postgresql.org/docs/14/spi.html)), do you have any related documentation at hand to share me?

# Context:

## Versions and plugins:
- psql 13.8
- timescaledb 2.8.0
- influx line listener 0.3 (installed from source)

## Regression tests:
```
bash-5.1# make installcheck
/usr/local/lib/postgresql/pgxs/src/makefiles/../../src/test/regress/pg_regress --inputdir=./ --bindir='/usr/local/bin'    --load-extension=influx --user=postgres --dbname=contrib_regression parse worker inval
(using postmaster on Unix socket, default port)
============== dropping database "contrib_regression" ==============
DROP DATABASE
============== creating database "contrib_regression" ==============
CREATE DATABASE
ALTER DATABASE
============== installing influx                      ==============
CREATE EXTENSION
============== running regression test queries        ==============
test parse                        ... ok           75 ms
test worker                       ... FAILED     3159 ms
test inval                        ... FAILED     4265 ms

======================
 2 of 3 tests failed. 
======================

The differences that caused some tests to fail can be viewed in the
file "/pg_influx/regression.diffs".  A copy of the test summary that you see
above is saved in the file "/pg_influx/regression.out".

make: *** [/usr/local/lib/postgresql/pgxs/src/makefiles/pgxs.mk:420: installcheck] Error 1

```

## Regression tests diff:
```
bash-5.1# cat /pg_influx/regression.diffs
diff -U3 /pg_influx/expected/worker.out /pg_influx/results/worker.out
--- /pg_influx/expected/worker.out
+++ /pg_influx/results/worker.out
@@ -21,48 +21,13 @@
 pg_sleep | 
 
 SELECT * FROM db_worker.cpu;
--[ RECORD 1 ]-----------------------------------------------------------------------
-_time   | Mon Nov 25 23:39:14 2019 PST
-_tags   | {"cpu": "cpu0", "host": "fury"}
-_fields | {"usage_user": "2.0408163264927324", "usage_system": "2.0408163264927324"}
--[ RECORD 2 ]-----------------------------------------------------------------------
-_time   | Mon Nov 25 23:39:14 2019 PST
-_tags   | {"cpu": "cpu1", "host": "fury"}
-_fields | {"usage_user": "3.9215686274649673", "usage_system": "3.921568627286635"}
+(0 rows)
 
 SELECT * FROM db_worker.disk;
--[ RECORD 1 ]-----------------------------------------------------------------------------------------------------------
-_time   | Mon Nov 25 23:39:14 2019 PST
-host    | fury
-device  | nvme0n1p2
-_tags   | {"mode": "rw", "path": "/", "fstype": "ext4"}
-_fields | {"free": "912578965504", "used": "42751348736", "total": "1006530654208", "used_percent": "4.475033200428718"}
--[ RECORD 2 ]-----------------------------------------------------------------------------------------------------------
-_time   | Mon Nov 25 23:39:14 2019 PST
-host    | fury
-device  | nvme0n1p1
-_tags   | {"mode": "rw", "path": "/boot/efi", "fstype": "vfat"}
-_fields | {"free": "527806464", "used": "7999488", "total": "535805952", "used_percent": "1.4929822952022749"}
+(0 rows)
 
 SELECT * FROM db_worker.system;
--[ RECORD 1 ]--------------------------------------------------------------------------------
-_time   | Tue Nov 26 07:39:14 2019
-host    | fury
-uptime  | 
-_tags   | {}
-_fields | {"load1": "2.13", "load5": "1.18", "load15": "0.84", "n_cpus": "8", "n_users": "1"}
--[ RECORD 2 ]--------------------------------------------------------------------------------
-_time   | Tue Nov 26 07:39:14 2019
-host    | fury
-uptime  | 607641
-_tags   | {}
-_fields | {}
--[ RECORD 3 ]--------------------------------------------------------------------------------
-_time   | Tue Nov 26 07:39:14 2019
-host    | fury
-uptime  | 
-_tags   | {}
-_fields | {"uptime_format": "7 days,  0:47"}
+(0 rows)
 
 SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE backend_type like '%Influx%';
 -[ RECORD 1 ]--------+--
diff -U3 /pg_influx/expected/inval.out /pg_influx/results/inval.out
--- /pg_influx/expected/inval.out
+++ /pg_influx/results/inval.out
@@ -13,10 +13,7 @@
 pg_sleep | 
 
 SELECT * FROM db_worker.cpu;
--[ RECORD 1 ]-------------------------------------------
-_time   | Mon Nov 25 23:39:14 2019 PST
-_tags   | {"cpu": "cpu0", "host": "fury"}
-_fields | {"usage_user": "5.40", "usage_system": "2.04"}
+(0 rows)
 
 ALTER TABLE db_worker.cpu ADD COLUMN cpu text;
 CALL send_packet('cpu,cpu=cpu0,host=fury usage_system=2.04,usage_user=5.40 1574753955000000000', 4711::text);
@@ -25,16 +22,7 @@
 pg_sleep | 
 
 SELECT * FROM db_worker.cpu;
--[ RECORD 1 ]-------------------------------------------
-_time   | Mon Nov 25 23:39:14 2019 PST
-_tags   | {"cpu": "cpu0", "host": "fury"}
-_fields | {"usage_user": "5.40", "usage_system": "2.04"}
-cpu     | 
--[ RECORD 2 ]-------------------------------------------
-_time   | Mon Nov 25 23:39:15 2019 PST
-_tags   | {"host": "fury"}
-_fields | {"usage_user": "5.40", "usage_system": "2.04"}
-cpu     | cpu0
+(0 rows)
 
 ALTER TABLE db_worker.cpu ADD COLUMN host text;
 CALL send_packet('cpu,cpu=cpu0,host=fury usage_system=2.04,usage_user=5.40 1574753956000000000', 4711::text);
@@ -43,24 +31,7 @@
 pg_sleep | 
 
 SELECT * FROM db_worker.cpu;
--[ RECORD 1 ]-------------------------------------------
-_time   | Mon Nov 25 23:39:14 2019 PST
-_tags   | {"cpu": "cpu0", "host": "fury"}
-_fields | {"usage_user": "5.40", "usage_system": "2.04"}
-cpu     | 
-host    | 
--[ RECORD 2 ]-------------------------------------------
-_time   | Mon Nov 25 23:39:15 2019 PST
-_tags   | {"host": "fury"}
-_fields | {"usage_user": "5.40", "usage_system": "2.04"}
-cpu     | cpu0
-host    | 
--[ RECORD 3 ]-------------------------------------------
-_time   | Mon Nov 25 23:39:16 2019 PST
-_tags   | {}
-_fields | {"usage_user": "5.40", "usage_system": "2.04"}
-cpu     | cpu0
-host    | fury
+(0 rows)
 
 SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE backend_type like '%Influx%';
 -[ RECORD 1 ]--------+--

```

--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,7 @@
 EXTENSION = influx
 DATA = influx--0.3.sql
 MODULE_big = influx
-OBJS = influx.o worker.o network.o parser.o cache.o metric.o
+OBJS = influx.o worker.o network.o ingest.o cache.o metric.o
 
 REGRESS = parse worker inval
 REGRESS_OPTS += --load-extension=influx
@@ -25,7 +25,7 @@ PGXS := $(shell $(PG_CONFIG) --pgxs)
 include $(PGXS)
 
 cache.o: cache.c cache.h
-influx.o: influx.c influx.h parser.h
+influx.o: influx.c influx.h ingest.h
 network.o: network.c network.h
-parser.o: parser.c parser.h
-worker.o: worker.c worker.h cache.h influx.h parser.h network.h
+ingest.o: ingest.c ingest.h
+worker.o: worker.c worker.h cache.h influx.h ingest.h network.h

--- a/influx.c
+++ b/influx.c
@@ -30,7 +30,7 @@
 #include <stdbool.h>
 #include <string.h>
 
-#include "parser.h"
+#include "ingest.h"
 
 PG_MODULE_MAGIC;
 
@@ -39,9 +39,9 @@ PG_FUNCTION_INFO_V1(parse_influx);
 /**
  * Parser state setup.
  */
-ParseState *ParseInfluxSetup(char *buffer) {
-  ParseState *state = palloc(sizeof(ParseState));
-  ParseStateInit(state, buffer);
+IngestState *ParseInfluxSetup(char *buffer) {
+  IngestState *state = palloc(sizeof(IngestState));
+  IngestStateInit(state, buffer);
   return state;
 }
 
@@ -58,7 +58,7 @@ ParseState *ParseInfluxSetup(char *buffer) {
  * @param tupdesc Tuple descriptor for the data.
  * @returns tuple Heap tuple
  */
-static HeapTuple ParseInfluxNextTuple(ParseState *state,
+static HeapTuple ParseInfluxNextTuple(IngestState *state,
                                       AttInMetadata *attinmeta) {
   int metric_attnum;
   Datum *values;
@@ -73,7 +73,7 @@ static HeapTuple ParseInfluxNextTuple(ParseState *state,
   /* Read lines until we find one that can be used. If none are found, we're
    * done. */
   do {
-    if (!ReadNextLine(state))
+    if (!IngestReadNextLine(state))
       return NULL;
   } while (!CollectValues(&state->metric, attinmeta, argtypes, values, nulls));
 
@@ -94,7 +94,7 @@ static HeapTuple ParseInfluxNextTuple(ParseState *state,
 Datum parse_influx(PG_FUNCTION_ARGS) {
   HeapTuple tuple;
   FuncCallContext *funcctx;
-  ParseState *state;
+  IngestState *state;
 
   if (SRF_IS_FIRSTCALL()) {
     TupleDesc tupdesc;

--- a/influx.h
+++ b/influx.h
@@ -27,9 +27,9 @@
 
 #include <stdbool.h>
 
-#include "parser.h"
+#include "ingest.h"
 
-ParseState *ParseInfluxSetup(char *buffer);
+IngestState *ParseInfluxSetup(char *buffer);
 Jsonb *BuildJsonObject(List *items);
 
 #endif /* INFLUX_H_ */

--- a/ingest.c
+++ b/ingest.c
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#include "parser.h"
+#include "ingest.h"
 
 #include <postgres.h>
 
@@ -46,7 +46,7 @@ static bool isident(char ch) {
  * @retval true the next character matches `ch`
  * @retval false The next character does not match `ch`
  */
-static bool CheckNextChar(ParseState *state, char ch) {
+static bool CheckNextChar(IngestState *state, char ch) {
   if (*state->current != ch)
     return false;
   *state->current++ = '\0';
@@ -62,7 +62,7 @@ static bool CheckNextChar(ParseState *state, char ch) {
  * @param state Parser state
  * @param ch Character to match
  */
-static void ExpectNextChar(ParseState *state, char ch) {
+static void ExpectNextChar(IngestState *state, char ch) {
   if (*state->current != ch)
     ereport(ERROR,
             (errcode(ERRCODE_SYNTAX_ERROR), errmsg("unexpected character"),
@@ -75,7 +75,7 @@ static void ExpectNextChar(ParseState *state, char ch) {
 /**
  * Initialize parser state.
  */
-void ParseStateInit(ParseState *state, char *line) {
+void IngestStateInit(IngestState *state, char *line) {
   memset(state, 0, sizeof(*state));
   state->current = line;
   state->start = line;
@@ -84,7 +84,7 @@ void ParseStateInit(ParseState *state, char *line) {
 /**
  * Read identifier from buffer.
  */
-static char *ReadIdent(ParseState *state) {
+static char *ReadIdent(IngestState *state) {
   char *begin = state->current;
   if (!*state->current)
     return NULL;
@@ -119,7 +119,7 @@ static char *ReadIdent(ParseState *state) {
  * @returns A pointer to the begining of the string, or null if
  * nothing was read.
  */
-static char *ReadValue(ParseState *state, Type *ptype) {
+static char *ReadValue(IngestState *state, Type *ptype) {
   char *const begin = state->current;
   char *rptr = state->current; /* Read pointer */
   char *wptr = state->current; /* Write pointer */
@@ -200,7 +200,7 @@ static char *ReadValue(ParseState *state, Type *ptype) {
  *
  * @returns NULL if there are no more items, pointer to an item otherwise.
  */
-static KVItem *ReadItem(ParseState *state, bool typed) {
+static KVItem *ReadItem(IngestState *state, bool typed) {
   KVItem *item = palloc0(sizeof(KVItem));
   item->key = ReadIdent(state);
   ExpectNextChar(state, '=');
@@ -220,7 +220,7 @@ static KVItem *ReadItem(ParseState *state, bool typed) {
  *
  * @returns The list of items
  */
-static List *ReadItemList(ParseState *state, bool typed) {
+static List *ParserReadItemList(IngestState *state, bool typed) {
   List *items = NIL;
   items = lappend(items, ReadItem(state, typed));
   while (CheckNextChar(state, ','))
@@ -238,7 +238,7 @@ static List *ReadItemList(ParseState *state, bool typed) {
  * @param state Parser state
  * @return true on success, false on end of file.
  */
-bool ReadNextLine(ParseState *state) {
+bool IngestReadNextLine(IngestState *state) {
   char *name;
   state->metric.timestamp = NULL;
   state->metric.tags = NIL;
@@ -249,9 +249,9 @@ bool ReadNextLine(ParseState *state) {
     return false;
   state->metric.name = name;
   if (CheckNextChar(state, ','))
-    state->metric.tags = ReadItemList(state, false);
+    state->metric.tags = ParserReadItemList(state, false);
   ExpectNextChar(state, ' ');
-  state->metric.fields = ReadItemList(state, true);
+  state->metric.fields = ParserReadItemList(state, true);
   ExpectNextChar(state, ' ');
   state->metric.timestamp = ReadValue(state, NULL);
   CheckNextChar(state, '\n');

--- a/ingest.h
+++ b/ingest.h
@@ -14,8 +14,8 @@
  * limitations under the License.
  */
 
-#ifndef PARSER_H_
-#define PARSER_H_
+#ifndef INGEST_H_
+#define INGEST_H_
 
 #include <postgres.h>
 
@@ -28,22 +28,23 @@
 #include "metric.h"
 
 /**
- * Parser state.
+ * Ingest parser state.
  *
  * @note All pointers in the parser state will point into the line
  * buffer.
  */
-typedef struct ParseState {
-  char *start; /**< Start of line buffer */
+typedef struct IngestState {
+  /** Start of line buffer */
+  char *start;
 
   /** Current read position of line buffer  */
   char *current;
 
   /** Metric resulting from the parse. */
   Metric metric;
-} ParseState;
+} IngestState;
 
-void ParseStateInit(ParseState *state, char *line);
-bool ReadNextLine(ParseState *state);
+void IngestStateInit(IngestState *state, char *line);
+bool IngestReadNextLine(IngestState *state);
 
-#endif /* PARSER_H_ */
+#endif /* INGEST_H_ */

--- a/worker.c
+++ b/worker.c
@@ -64,13 +64,13 @@ static volatile sig_atomic_t ShutdownWorker = false;
  * Process one packet of lines.
  */
 static void ProcessPacket(char *buffer, size_t bytes, Oid nspid) {
-  ParseState *state;
+  IngestState *state;
 
   buffer[bytes] = '\0';
   state = ParseInfluxSetup(buffer);
 
   while (true) {
-    if (!ReadNextLine(state))
+    if (!IngestReadNextLine(state))
       return;
     MetricInsert(&state->metric, nspid);
   }

--- a/worker.c
+++ b/worker.c
@@ -158,6 +158,12 @@ void WorkerMain(Datum arg) {
                 errdetail("service=%s, database_id=%u, namespace_id=%u",
                           args->service, database_id, args->namespace_id)));
 
+  /* We need to start a transaction first because none is started and
+     SPI_connect_ext might use TopTransactionContext, which is set by
+     this function. The SPI_commit below will automatically start a
+     new one. */
+  StartTransactionCommand();
+
   /* This is the same approach as used in pgstats.c: We read the
      packets in two loops. One outer that will block when there is no
      data received, and one inner loop that will read packets as long
@@ -180,7 +186,6 @@ void WorkerMain(Datum arg) {
        sure that it does not change while we are updating it. */
     if ((err = SPI_connect_ext(SPI_OPT_NONATOMIC)) != SPI_OK_CONNECT)
       elog(ERROR, "SPI_connect failed: %s", SPI_result_code_string(err));
-    SPI_start_transaction();
     PushActiveSnapshot(GetTransactionSnapshot());
 
     pgstat_report_activity(STATE_RUNNING, "processing incoming packets");

